clientv3/ordering/kv.go

```diff
@@ -0,0 +1,99 @@
+// Copyright 2017 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ordering
+
+import (
+	"context"
+	"sync/atomic"
+
+	"github.com/coreos/etcd/clientv3"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+)
+
+// kvOrdering guarantees that the revision in get responses
+// are equal to or greater than the stored previous known revision.
+type kvOrdering struct {
+	clientv3.KV
+	kvOrderViolation  KvOrderViolationFunc
+	txnOrderViolation TxnOrderViolationFunc
+	prevRev           int64
+}
+
+func NewKV(kv clientv3.KV, kvViolation KvOrderViolationFunc, txnViolation TxnOrderViolationFunc) *kvOrdering {
+	return &kvOrdering{kv, kvViolation, txnViolation, 0}
+}
+
+func (kv *kvOrdering) getPrevRev() int64 {
+	return atomic.LoadInt64(&kv.prevRev)
+}
+
+func (kv *kvOrdering) setPrevRev(currRev int64) {
+	atomic.StoreInt64(&kv.prevRev, currRev)
```

Comment:
Probably want to prevent this from going backwards if there are concurrent calls to `setPrevRev`:
```go
for { 
    pr := kv.getPrevRev()
    if currentRev <= pr {
        return
    }
    if atomic.CompareAndSwapInt64(&kv.prevRev, pr, currRev) {
        return
    }
}